lib/solarwinds_apm/opentelemetry/otlp_processor.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+module SolarWindsAPM
+  module OpenTelemetry
+    # reference: OpenTelemetry::SDK::Trace::SpanProcessor; inheritance: SolarWindsProcessor
+    class OTLPProcessor < SolarWindsProcessor
+      attr_accessor :description
+
+      # @param [Hash] meters the hash of meter created by ::OpenTelemetry.meter_provider.meter('meter_name')
+      # @param [TxnNameManager] txn_manager storage for transaction name
+      # @exporter [Exporter] exporter reporter that send trace data
+      def initialize(exporter)
+        super(exporter, nil)
+        @meters   = init_meters
+        @metrics  = {}
+      end
+
+      # @param [Span] span the {Span} that just started.
+      # @param [Context] parent_context the
+      #  started span.
+      def on_start(span, parent_context)
+        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] processor on_start span: #{span.inspect}" }
+
+        initialize_metrics if @metrics.empty?
+
+        parent_span = ::OpenTelemetry::Trace.current_span(parent_context)
+        return if parent_span && parent_span.context != ::OpenTelemetry::Trace::SpanContext::INVALID && parent_span.context.remote? == false
+
+        span_attrs = span_attributes(span)
+        span.add_attributes(span_attrs)
+      rescue StandardError => e
+        SolarWindsAPM.logger.info { "[#{self.class}/#{__method__}] processor on_start error: #{e.message}" }
+      end
+
+      # @param [Span] span the {Span} that just ended.
+      def on_finish(span)
+        SolarWindsAPM.logger.debug { "[#{self.class}/#{__method__}] processor on_finish span: #{span.to_span_data.inspect}" }
+
+        # metrics per trace, therefore, we only record the parent span (span.parent_span_id has to be 00000000000 INVALID_SPAN_ID to qualify as parent span)
+        if span.parent_span_id != ::OpenTelemetry::Trace::INVALID_SPAN_ID
+          return unless span.context.trace_flags.sampled?
+
+          @exporter&.export([span.to_span_data])
+          record_sampling_metrics
+          ::OpenTelemetry.meter_provider.metric_readers.each(&:pull)
+        end
```

Comment:
I see, the crux seems you're implementing the span processor (rather than extending or simply using the upstream batch or simple processor implementations, which is what our other APM custom distros do). This works, but the downside is it can't take advantage of the efficient batching and export already implemented by upstream--currently our [SolarWindsProcessor](https://github.com/solarwinds/apm-ruby/blob/main/lib/solarwinds_apm/opentelemetry/solarwinds_processor.rb) does the "simple processor" way of exporting per span.

Does OTel Ruby allow multiple span processors? That's how our other custom distros do it--conceptually like below:

SDK is configured with:
* a "metrics" processor that simply records metrics (w/o exporting any spans)
* an upstream span processor that actually takes in an exporter (whether our custom apm-proto, or standard OTLP) that does actual span exports

The two lower diagrams in [this section](https://swicloud.atlassian.net/wiki/spaces/NIT/pages/3306226004/Span+Processor#What-does-it-do%3F) illustrate the point, so does our python custom distro config:
* sets up various processors that deal with [transaction naming and recording metrics](https://github.com/solarwinds/apm-python/blob/main/solarwinds_apm/configurator.py#L139-L149)
* sets up upstream simple/batch processor [for actual span export](https://github.com/solarwinds/apm-python/blob/main/solarwinds_apm/configurator.py#L357-L370)

Our node.js custom distro came up with a [CompoundSpanProcessor](https://github.com/solarwinds/apm-js/blob/d67857535d8995b1ccd8ae3483208c66ebe3e883/packages/sdk/src/compound-processor.ts#L26) to essentially support the same idea, note that it extends the BatchSpanProcessor to leverage upstream implementation of batched span exports, while each of the processors [in the compound list](https://github.com/solarwinds/apm-js/blob/d67857535d8995b1ccd8ae3483208c66ebe3e883/packages/solarwinds-apm/src/init.ts#L286) extend the noop processor and just [override the needed parts](https://github.com/solarwinds/apm-js/blob/d67857535d8995b1ccd8ae3483208c66ebe3e883/packages/sdk/src/response-time-processor.ts#L32).

How about restructuring ruby custom distro to do something similar?

Reply:
Thanks for the explanation; I have updated the our agent configuration to include multiple processors that do span and metrics exporting separately!